### PR TITLE
Qgisserveroutput

### PR DESCRIPTION
**Pull request by rduivenvoorde**
As promised the first request is to make it for (ignorant) users like me more obvious which port qgis_mapserver is running.

Then I thought to add some more info in the server responses. In my (and some people I know) experience trying to setup qgisserver for the first time is sometimes harder that it should be.
Because QGIS itself (I know in the logs it often is) does not tell you if there was just one layer wrong, or if there was an reading issue with a project file (one of the things I often hit on Windows servers) etc

So I added some more commits, though I can also imagine that it was a deliberate choice to NOT tell users what was wrong with configuration of a server... 
But... if you ask me: we should tell the (early adopting) users...
An alternative would be to run the server in some kind of development or production mode?

### CODE REVIEW

File: src/server/qgsserver.cpp
Comment (elpaso) [inline src/server/qgsserver.cpp:400]: Maybe also specify that the project can be set from the command line with `-p <path to project>`

File: src/server/qgsserver.cpp
Comment (elpaso) [inline src/server/qgsserver.cpp:418]: You can probably dynamically get the list of available service and APIs directly from the server, it would be safer in case the user added custom services or APIs through plugins or in case we add new core ones in the future.

File: src/server/qgsconfigcache.cpp
Comment (elpaso) [inline src/server/qgsconfigcache.cpp:96]: No: this is disclosure of path informations.

**Review by elpaso — CHANGES_REQUESTED** [review]

File: src/server/qgsserver.cpp
Comment (elpaso) [reply src/server/qgsserver.cpp:400]: Forget my comment. I thought we were still in the development server code part.

File: src/server/qgsserver.cpp
Comment (rduivenvoorde) [reply src/server/qgsserver.cpp:418]: @elpaso yes, I already found: 
server.sServiceRegistry.mServiceVersions
server.sServiceRegistry.mApiVersions
but before putting to much effort in it I first thought to wait for you to see IF you want all this or not :-) I'll happy to try.

File: src/server/qgsconfigcache.cpp
Comment (rduivenvoorde) [reply src/server/qgsconfigcache.cpp:96]: Ok, thought about this too. And tried to find out if I could only show the project filename... 
As said: trying to setup mapserver in osgeo4w  on windows servers with network paths' it was often hard to find what was the problem: permissions, the projectfile or something IN the projectfile..
So would only show the project file name be ok for you?

Comment (elpaso) [reply src/server/qgsconfigcache.cpp:96]: Well, I would advice against it, if for some reason a rewrite rule was in place in order to hide the project file name we shouldn't disclose it. 

But I don't really understand why if (as you say) you are trying to ease the setup process you  don't just point the user to the server logs, that contain the full path and some more information about the various issues.

File: src/server/qgsserver.cpp
Comment (elpaso) [reply src/server/qgsserver.cpp:418]: I think that listing the supported services and APIs is a good idea.
CC @pblottiere @rldhont what do you think?

Comment (pblottiere) [reply src/server/qgsserver.cpp:418]: > I think that listing the supported services and APIs is a good idea.

+1

Thanks for taking a look @rduivenvoorde :)

Comment (rduivenvoorde) [reply src/server/qgsserver.cpp:418]: @elpaso @pblottiere  why is QGIS so strict about the service parameter case?
'WMS' is ok, but 'wms' fails with (now) error above?
Should we not just cast to uppercase and check for 'WMS'? 
(I know the specs state: keys are case-INsensitive, and value are case-sensitive), but in this case both could/should work (like in all other WMS's I wordt with?)

File: src/server/qgsconfigcache.cpp
Comment (rduivenvoorde) [reply src/server/qgsconfigcache.cpp:96]: Because (well in my experience) it is not so easy to set up the right logging on Windows when you are  not mixed Linux/Windows admin, trying to set Apache env params on a Windows machine etc etc :-)
But as said: one of my clients is bitten by this pretty often because of permission errors on the network drives where either the project files OR the data files OR they changed the db-parameters  etc etc

Comment (elpaso) [reply src/server/qgsconfigcache.cpp:96]: Sorry, but there is no way we can disclose this kind of information to the world within the server HTLM response.

File: src/server/qgsserver.cpp
Comment (rldhont) [reply src/server/qgsserver.cpp:418]: > I think that listing the supported services and APIs is a good idea.

+1

Comment (pblottiere) [reply src/server/qgsserver.cpp:418]: > why is QGIS so strict about the service parameter case?
'WMS' is ok, but 'wms' fails with (now) error above?

It's a good question actually :). It should be checked but I don't think it's an OGC curb...

**Comment by rduivenvoorde** [comment]
@elpaso I reverted that last commit. Is it ok like this? 
Or should I try to "dynamically get the list of available services"?
I could 'squash' all this to 1 commit too?

**Review by elpaso — APPROVED** [review]

**Comment by elpaso** [comment]
> @elpaso I reverted that last commit. Is it ok like this?

Yes, thank you!

> Or should I try to "dynamically get the list of available services"?

That is up to you, I think we all agree that it would be a nice little enhancement but it can also land to a separate PR.

> I could 'squash' all this to 1 commit too?

I can merge & squash from the GUI here.

**Comment by elpaso** [comment]
@rduivenvoorde can you please fix the test control files so we can merge?

**Comment by rduivenvoorde** [comment]
@elpaso ah duh, so tests are failing because of my edits :-) I will try to fix. I was under the impression that Travis always fails every now and then, and I planned to ask you how to 'rerun' a travis run....
But now I will fix the test, and sorry about making fun of Travis ;-)

**Comment by rduivenvoorde** [comment]
Ai, failing again on Travis. Thought I fixed the python tests...
I tried to run 'make test', but  that resulted in a more then 1 hour run, and A LOT of failing tests.
@elpaso is there a quick/easy way to see what still is wrong? 
Tried `ctest -R serv -V` and `ctest -R wms -V` those were OK.
And `python3 ../tests/src/python/test_qgsserver.py` also was fine...
The travis error is in PyQgsServerWMSGetPrint and indeed `ctest -R PyQgsServerWMSGetPrint -V` fails with me also. But on image related errors...?

**Comment by rldhont** [comment]
@rduivenvoorde about `PyQgsServerWMSGetPrint` I have open an issue https://github.com/qgis/QGIS/issues/35973 after these PR https://github.com/qgis/QGIS/pull/35277 https://github.com/qgis/QGIS/pull/35937

**Comment by rduivenvoorde** [comment]
@rldhont  mmm, thought to rebase but apperently I messed up. :-(

Probably best to close this and redo my edits in a new PR?

**Comment by rldhont** [comment]
@rduivenvoorde it seems you made a merge and not a rebase.

**Comment by rldhont** [comment]
You can 
```
git checkout qgisserveroutput
#rename your local branch
git branch -m qgisserveroutput old
# go to master
git checkout master
# update master
git pull upstream master
# new branch
git checkout -b qgisserveroutput
# cherry picking your commits
git cherry-pick cce8321 
git cherry-pick fd5f98f
git cherry-pick ac48c0d
git cherry-pick 0d7f6d2
git cherry-pick 5da32d0 
git cherry-pick 98db81e 
git cherry-pick 5caba74 
# update your remote branch
git push origin qgisserveroutput -f
```

**Comment by rduivenvoorde** [comment]
> You can
> 
> ```
> git checkout qgisserveroutput
> #rename your local branch
> git branch -m qgisserveroutput old
> # go to master
> git checkout master
> # update master
> git pull upstream master
> # new branch
> git checkout -b qgisserveroutput
> # cherry picking your commits
> git cherry-pick cce8321 
> git cherry-pick fd5f98f
> git cherry-pick ac48c0d
> git cherry-pick 0d7f6d2
> git cherry-pick 5da32d0 
> git cherry-pick 98db81e 
> git cherry-pick 5caba74 
> # update your remote branch
> git push origin qgisserveroutput -f
> ```

Thanks René-Luc !! Done.. See if all is green now...

File: tests/src/python/test_qgsserver.py
Comment (elpaso) [inline tests/src/python/test_qgsserver.py:333]: Actually I'm not fully convinced about this. I mean, I can live with that, but exposing `MAP` in the query string URL is not only bad practice but it is also violating WFS3 specs (I know this code is not going to be hit by WFS3, but still).

So, if you really wanted to be precise you should list the other preferred methods that are available to make QGIS aware of the project it needs to serve, but that is probably too much for an error message.

**Review by elpaso — APPROVED** [review]

File: tests/src/python/test_qgsserver.py
Comment (rduivenvoorde) [reply tests/src/python/test_qgsserver.py:333]: I agree, it's just that: telling people that there is a "Project file error" if they are trying to get the server working seems not appropriate. If they add a MAP and SERVICE param, it will give them futher hints...

Comment (elpaso) [reply tests/src/python/test_qgsserver.py:333]: It depends who "people" are. 
Who looks at the XML output of a machine 2 machine designed service must be a special kind of people that should also be prepared to look at the server logs and RTFM.

The server is **not** an end user application (not yet).

IMHO a better message would be: "A valid QGIS project could not be loaded. Please check server configuration." 

Because even "file" is wrong wording: what if the server is configured to use projects stored indide a DB?

Comment (rduivenvoorde) [reply tests/src/python/test_qgsserver.py:333]: Hi Gents, never mind, I think we just differ in what a server should and should not tell the user. 

QGIS in my opinion hides everything for the average user. Other WMS's like Minnesota Mapserver and Geoserver tell the user if they have configured something wrong.

I think OGC servers are not only there to be used for machine<->machine communication. It all starts with average users who try to setup servers and fail.

I've tried and failed :-)

Comment (elpaso) [reply tests/src/python/test_qgsserver.py:333]: @rduivenvoorde believe me when I say that I understand your frustration, I get my daily share of it.

In this PR there is value (I would have already merged the logging order change for instance).

But if the current message "Project file error" is wrong, replacing it with another slightly better but still wrong message is not what we want.

Would it be ok for you to change it to something that is more descriptive but also correct?

What comes to my mind is:
- A valid QGIS project could not be loaded. This may be the result of an invalid project or a missing project identifier, please refer to the QGIS Server logs and the official server documentation for further details.

Also note that when the project is valid and the `SERVICE` is missing you get a different error: 
```xml
<ServiceException code="Service configuration error">Service unknown or unsupported</ServiceException>
```
Which is a 200OK response and not (like in the previous case) an internal server error with a 500 code.

And I insist: if the average user (whoever she is) is trying to set up a web server she must be prepared to look at the server logs (where there are much more details about the error).

Comment (rduivenvoorde) [reply tests/src/python/test_qgsserver.py:333]: @elpaso it all started with this change of order, which was easy. All other stuff took me considerable (learning) time, for example to find out how to run the tests (still not able to run them in Qt-Creator). You are probably right, but it takes me too much time already. Maybe just cherry-pick the first commit and leave rest as is?

Comment (rldhont) [reply tests/src/python/test_qgsserver.py:333]: thanks @rduivenvoorde 
I'll do the update

Comment (elpaso) [reply tests/src/python/test_qgsserver.py:333]: > @elpaso it all started with this change of order, which was easy. All other stuff took me considerable (learning) time, for example to find out how to run the tests (still not able to run them in Qt-Creator). 

That might be tricky, I can attach my config as an example:

immagine

And my env:
```
LD_LIBRARY_PATH=/home/xxx/local5_13_2_release_gdal3/lib/:/home/xxx/dev/oracle/instantclient_19_3
OGR_GPKG_FOREIGN_KEY_CHECK=NO
PYTHONPATH=/home/xxx/local5_13_2_release_gdal3/lib/python3/dist-packages:/home/xxx/dev/QGIS/build-local-qt5132-release_gdal3/output/python/:/home/xxx/dev/QGIS/build-local-qt5132-release_gdal3/output/python/plugins:/home/xxx/dev/QGIS/tests/src/python/:/home/xxx/local13_2_release_gdal3/lib/python3/dist-packages/:/home/xxx/.local/lib/python3.6/site-packages/:/usr/lib/python3.6:/usr/lib/python3.6/plat-x86_64-linux-gnu:/usr/lib/python3.6/lib-dynload:/home/xxx/.local/lib/python3.6/site-packages:/usr/local/lib/python3.6/dist-packages:
QGIS_PREFIX_PATH=/home/xxx/dev/QGIS/build-local-qt5132-release_gdal3/output
```

> You are probably right, but it takes me too much time already. Maybe just cherry-pick the first commit and leave rest as is?

Whatever works for you, I'll approve (actually I had already approved this one, I just hoped for a better error message).

Comment (rduivenvoorde) [reply tests/src/python/test_qgsserver.py:333]: @elpaso thanks. Actually I do not know who is doing the merge after an approval, I can do it myself, as I have commit rights, but I thought that a select group of core committers actually did all the merges... Was I supposed to do it myself?

Comment (elpaso) [reply tests/src/python/test_qgsserver.py:333]: Once it's approved you can merge it by yourself, but in this case this PR has been already merged by @rldhont 

Thank you for your patience :)

Comment (rldhont) [reply tests/src/python/test_qgsserver.py:333]: @rduivenvoorde thanks for your work, I have merge it because it has been approved by @elpaso.

**Pull request merged**
